src/app/(proper_react)/redesign/(authenticated)/user/(dashboard)/settings/actions.ts

```diff
@@ -0,0 +1,153 @@
+    session.user.email,
+  )) as SubscriberRow | null;
+  if (!subscriber) {
+    return {
+      success: false,
+      error: "no-subscriber-data-found",
+      errorMessage: l10n.getString("user-add-invalid-email"),
+    };
+  }
+  const submittedAddress = formData.get("newEmailAddress");
+  const validatedEmailAddress =
+    typeof submittedAddress === "string"
+      ? validateEmailAddress(submittedAddress)
+      : null;
+
+  if (validatedEmailAddress === null) {
+    return {
+      success: false,
+      error: "invalid-email",
+      errorMessage: l10n.getString("user-add-invalid-email"),
+    };
+  }
+
+  const existingAddresses = [session.user.email]
+    .concat(subscriber.email_addresses.map((emailRow) => emailRow.email))
+    .map((address) => address.toLowerCase());
+  if (
+    existingAddresses.length >=
+    Number.parseInt(process.env.NEXT_PUBLIC_MAX_NUM_ADDRESSES!, 10)
+  ) {
+    return {
+      success: false,
+      error: "too-many-emails",
+      errorMessage: l10n.getString("user-add-too-many-emails"),
+    };
+  }
+
+  if (existingAddresses.includes(validatedEmailAddress.email.toLowerCase())) {
+    return {
+      success: false,
+      error: "address-already-added",
+      errorMessage: l10n.getString("user-add-duplicate-email"),
+    };
+  }
+
+  try {
+    const unverifiedSubscriber = await addSubscriberUnverifiedEmailHash(
+      subscriber,
+      validatedEmailAddress.email,
+    );
+
+    await initEmail();
+    await sendVerificationEmail(
+      // This first parameter is unused, so the type assertion is safe.
+      // When non-TS invocations of this function are removed, we should remove
+      // the unused parameter:
+      session.user.subscriber as unknown as Parameters<
+        typeof sendVerificationEmail
+      >[0],
+      unverifiedSubscriber.id,
+      getL10n(),
+    );
+    revalidatePath("/redesign/user/settings/");
+
+    return {
+      success: true,
+      submittedAddress: validatedEmailAddress.email,
+    };
+  } catch (e) {
+    if (e instanceof Error && e.message === "error-email-validation-pending") {
+      return {
+        success: false,
+        error: "verification-email-just-sent",
+        errorMessage: l10n.getString("user-add-verification-email-just-sent"),
+      };
+    }
+
+    return {
+      success: false,
+      error: "unknown-error",
+      errorMessage: l10n.getString("user-add-unknown-error"),
+    };
+  }
+}
+
+export async function onRemoveEmail(email: EmailRow) {
+  const session = await getServerSession();
+  if (!session?.user.email) {
+    logger.error(
+      `Tried to delete email [${email.id}] without an active session.`,
+    );
+    return;
+  }
+  const subscriber = (await getSubscriberByEmail(
+    session.user.email,
+  )) as SubscriberRow | null;
+  if (email.subscriber_id !== subscriber?.id) {
+    logger.error(
+      `Subscriber [${subscriber?.id}] tried to delete email [${email.id}], which belongs to another subscriber.`,
```

Comment:
Question: We are only logging this on the server so it does not matter as much, but wondering in general if it might be preferable to try and not log conclusions that relate to other user info and rather say: “…, which does not belong to the current subscriber.”?

Reply:
Yes, I considered that too (and it is relevant on the server too, since the people who can access logs are not the same as the people who can access the database). In this case, however, I figured that this attempt is likely to be malicious, and thus warrants the increased detail? /cc @rhelmer